src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,136 @@
+            throw new ArrayListIndexOutOfBoundsException(
+                    String.format("Array index %d out of bounds", index)
+            );
+        }
+    }
+
+    private void rangeCheckForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    String.format("Index %d not suitable for add", index)
+            );
+        }
+    }
+
+    private int indexOf(Object o) {
+        int index = 0;
+        boolean found = false;
+        if (o == null) {
+            for (; index < size; index++) {
+                if (elementData[index] == null) {
+                    found = true;
+                    break;
+                }
+            }
+        } else {
+            for (; index < size; index++) {
+                if (o.equals(elementData[index])) {
+                    found = true;
+                    break;
+                }
+            }
+        }
+        return found ? index : INDEX_NOT_FOUND;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size + 1 < elementData.length) {
+            elementData[size++] = value;
+            return;
+        }
+        grow();
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        rangeCheckForAdd(index);
+        if (size == 0 || index == size) {
+            add(value);
+            return;
+        }
+        final int newSize = size + 1;
+        final Object[] newArray = new Object[newSize];
+        System.arraycopy(elementData, index, newArray, index + 1, size - index);
+        System.arraycopy(elementData, 0, newArray, 0, index);
+        elementData = Arrays.copyOf(newArray, size = newSize);
+        elementData[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null || list.isEmpty()) {
+            return;
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elementData[index];
     }
 
     @Override
-    public void set(T value, int index) {
-
+    public void set(T value, int index) { //++
+        checkIndex(index);
+        elementData[index] = value;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedValue = (T) elementData[index];
+        final int newSize = size - 1;
+        if (newSize > index) {
+            System.arraycopy(elementData, index + 1, elementData, index, newSize - index);
+        }
+        elementData = Arrays.copyOf(elementData, size = newSize);
+        return removedValue;
     }
 
     @Override
-    public T remove(T element) {
-        return null;
+    public T remove(T element) { //++
```

Comment:
For what is this comment?

Reply:
nvm, its useless comments that i forget 2remove, removed.
+removed in new commit.